### PR TITLE
Fix 22384 - Add support for default values in C-style bitfields

### LINKED ISSUE #22384: Error: initializer not allowed for bitfield declaration
Given for example:
```d
struct S
{
    int a = 1;
    int b : 16;
    int c : 16;
    int d = 4;
}
```
The initializer symbol is emitted as:
```
	.long	1
	.value	0
	.value	0
	.long	4
```
It ought to be possible to specify the default value of bit fields as well.
```d
struct S
{
    int a = 1;       // 	.long	1
    int b : 16 = 2;  // 	.value	2
    int c : 16 = 3;  // 	.value	3
    int d = 4;       // 	.long	4
}
```
Fields smaller than a byte just get merged into the byte representation.
```d
struct T
{
    int a = 1;       // 	.long	1
    int b : 4 = 2;   // 	.byte	0x32
    int c : 4 = 3;
    int d : 8 = 4;   // 	.byte	0x4
    int e : 4 = 5;   // 	.byte	0x65
    int f : 12 = 6;  // 	.byte	0x0
    int g = 7;       // 	.long	7
}
```

### PR DESCRIPTION
**Pull request by MadhurKumar004**
This implements parsing, semantic analysis, and backend support for default initializers in bitfields.

- Modified 'declaration.d' & 'parse.d' to parse assignment expressions after bitfield width.
- Updated 'dsymbolsem.d' to validate initializers and check for overflows.
- Updated 'todt.d' to pack default values into the binary representation.
- Added regression tests.

### CODE REVIEW

File: compiler/test/runnable/test22384.d [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+T t;
+
+extern(C) int printf(const char*, ...);
+
+void main()
+{
+    if (s.a != 1) { printf("Error: s.a = %d\n", s.a); assert(0); }
+    if (s.b != 2) { printf("Error: s.b = %d\n", s.b); assert(0); }
+    if (s.c != 3) { printf("Error: s.c = %d\n", s.c); assert(0); }
+    if (s.d != 4) { printf("Error: s.d = %d\n", s.d); assert(0); }
```
Comment (ibuclaw) [inline compiler/test/runnable/test22384.d]:
```suggestion
    assert(s.a == 1);
    assert(s.b == 2);
    assert(s.c == 3);
    assert(s.d == 4);
```

File: compiler/src/dmd/glue/todt.d [hunk 2]
```diff
@@ -1008,7 +1008,23 @@ private void membersToDt(AggregateDeclaration ad, ref DtBuilder dtb,
...
-        else if (!bf)
+        else if (bf)
+        {
+            if (Initializer init = vd._init)
+            {
+                if (auto ei = init.isExpInitializer())
+                {
+                    if (auto ie = ei.exp.isIntegerExp())
+                    {
+                        auto value = ie.getInteger();
```
Comment (ibuclaw) [inline compiler/src/dmd/glue/todt.d]: I think it is OK to flatten this a bit.  Make it an ICE if codegen receives something that shouldn't have passed semantic without an error.


```suggestion
                if (init.isVoidInitializer())
                    continue;

                assert(vd.semanticRun >= PASS.semantic2done);

                auto ei = init.isExpInitializer();
                assert(ie);
                auto ie = ei.exp.isIntegerExp();
                assert(ie);
                auto value = ie.getInteger();
```

File: compiler/test/runnable/test22384.d [hunk 3]
```diff
@@ -0,0 +1,42 @@
+//https://github.com/dlang/dmd/issues/22384
+struct S
+{
+    int a = 1;
+    int b : 16 = 2;
```
Comment (ibuclaw) [inline compiler/test/runnable/test22384.d:5]: Do we need tests for mixed named and anonymous bitfields?

File: compiler/src/dmd/dsymbolsem.d [hunk 4]
```diff
@@ -3219,6 +3219,39 @@ private extern(C++) final class DsymbolSemanticVisitor : Visitor
             dsym.errors = true;
         }
         dsym.fieldWidth = cast(uint)uwidth;
+
+        if (dsym._init)
+        {
+            dsym._init = dsym._init.initializerSemantic(sc, dsym.type, INITinterpret);
+            if (auto ei = dsym._init.isExpInitializer())
```
Comment (ibuclaw) [inline compiler/src/dmd/dsymbolsem.d]: This should be moved to `dmd.semantic2`, which is the correct pass to do all checks on symbol initializers - and the initializerSemantic of the bitfield has already been ran at this point I think.

**Comment by ibuclaw** [comment]
@MadhurKumar004 thanks for picking this up! :+1:

File: compiler/src/dmd/dsymbolsem.d [hunk 5]
```diff
@@ -3219,6 +3219,39 @@ private extern(C++) final class DsymbolSemanticVisitor : Visitor
...
+            {
+                if (auto ie = ei.exp.isIntegerExp())
+                {
+                    auto value = ie.toInteger();
+                    if (uwidth < 64)
+                    {
+                        const mask = (1L << uwidth) - 1;
+                        if (dsym.type.isUnsigned())
+                        {
+                            if ((value & mask) != value)
```
Comment (ibuclaw) [inline compiler/src/dmd/dsymbolsem.d]: I think `dmd.intrange` would be better than doing this yourself.

Something like:
```d
auto value = getIntRange(ei.exp);
auto bounds = IntRange(SignExtendedNumber(dsym.getMinMax(Id.min), !dsym.type.isUnsigned),
                       SignExtendedNumber(dsym.getMinMax(Id.max)));

if (!bounds.contains(value))
{
    error(ei.loc, ...);
}
```

**Comment by MadhurKumar004** [comment]
@ibuclaw Made the changes as you suggested

File: compiler/src/dmd/semantic2.d [hunk 6]
```diff
@@ -320,6 +321,41 @@ private extern(C++) final class Semantic2Visitor : Visitor
...
+            return;
+
+        auto ei = bfd._init.isExpInitializer();
+        if (!ei)
+            return;
+
+        ei.exp = ei.exp.optimize(WANTvalue);
+        auto e = ei.exp;
+        if (!e.isIntegerExp())
+            return;
```
Comment (ibuclaw) [inline compiler/src/dmd/semantic2.d]: This is redundant. `visit(VarDeclaration)` has already ran either `ctfeInterpret()` or `optimize()` on the initializer expression, and issued an error if it can't.

File: compiler/src/dmd/semantic2.d [hunk 7]
```diff
@@ -320,6 +321,41 @@ private extern(C++) final class Semantic2Visitor : Visitor
...
+
+        import dmd.intrange;
+        auto value = getIntRange(e);
```
Comment (ibuclaw) [inline compiler/src/dmd/semantic2.d]: As per above comment about `optimize`.
```suggestion
        auto value = getIntRange(ei.exp);
```

File: compiler/src/dmd/semantic2.d [hunk 8]
```diff
@@ -320,6 +321,41 @@ private extern(C++) final class Semantic2Visitor : Visitor
...
+        const bool isUnsigned = bfd.type.isUnsigned();
+        auto bounds = IntRange(
+            SignExtendedNumber(bfd.getMinMax(Id.min), !isUnsigned),
+            SignExtendedNumber(bfd.getMinMax(Id.max), false)
+        );
+
+        if (!bounds.contains(value))
+        {
+            const uwidth = bfd.fieldWidth;
+            error(bfd.loc, "bitfield initializer `%s` does not fit in %d bit%s",
```
Comment (ibuclaw) [inline compiler/src/dmd/semantic2.d]: The error is about the bitfield _initializer_, but the location passed is the declaration.

I think you mean to error at the location of the expinitializer instead.
```suggestion
            error(ei.loc, "bitfield initializer `%s` does not fit in %d bit%s",
```

**Comment by MadhurKumar004** [comment]
@ibuclaw review please.

File: compiler/src/dmd/semantic2.d [hunk 9]
```diff
@@ -320,6 +321,39 @@ private extern(C++) final class Semantic2Visitor : Visitor
...
+
+        if (!bfd._init)
+            return;
+
+        auto ei = bfd._init.isExpInitializer();
+        if (!ei)
+            return;
+
+        if (!ei.exp.isIntegerExp())
+            return;
```
Comment (ibuclaw) [inline compiler/src/dmd/semantic2.d:340]: Open question, does this need more fail_compilation tests?

```d
float f : 0;
float f : 7;
float f : 7 = 4;
float f : 7 = 4.2f;
float f : 7 = cast(int)4.2;

int i : 0 = 1;
int i : 7 = 4.2f;

ubyte : 0 = 4;
ubyte : 8 = 4; 
ubyte : 8 = 4.2f;
```

What do these do?
```d
int i : 2 = true;
int i : 7 = 'A';
int i : 7 = cast(int)4.2;

char c : 4 = true;
char c : 8; // value is char.init or 0?

bool b : 1 = true;
bool b : 7 = 42;
```

If you can think of any other combinations of type/bit width/scalar initializer. Feel free to add it, though I'll let you use your judgement on whether it's necessary.

**Review by ibuclaw — COMMENTED** [review]
Can't comment on the backend part. Frontend looks ok, maybe just need more combinations of tests to ensure we don't start accepting invalid code.

If there's bitfield documentation on dlang.org, that requires updating too.

As this is a feature fix, would be best to include a changelog entry too.

Thanks!

**Comment by MadhurKumar004** [comment]
@ibuclaw Thanks for the advice on test combination. Because of that i was able to fix couple of edge cases that were giving ICE. Also added the changelog and currently working on the dlang.org documentation. Will send a pr for that after this get merged. Please check and review

File: changelog/dmd.bitfield-default-init.dd [hunk 10]
```diff
@@ -0,0 +1,16 @@
+Support for default values in C-style bitfields
+
+D now supports specifying default values for C-style bitfields in structs and classes.
+This brings valid D code closer to behavior allowed in C (via extensions) and allows for more concise initialization of bitfield members.
```
Comment (ibuclaw) [inline changelog/dmd.bitfield-default-init.dd]: Not sure about C, but it's as per C++20.

File: compiler/test/fail_compilation/fail22384.d [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+    float f : 0;        // Error: bitfield type `float` is not an integer type
+    float f2 : 7;       // Error: bitfield type `float` is not an integer type
+    float f3 : 7 = 4;   // Error: bitfield type `float` is not an integer type
+    float f4 : 7 = 4.2f;// Error: bitfield type `float` is not an integer type
+}
+
+struct InitFail {
+    int i : 7 = 4.2f;   // Error: cannot implicitly convert
+    // ubyte : 0 = 4;   // Error: zero width (tested above)
+    ubyte : 8 = 4.2f;   // Error: cannot implicitly convert
```
Comment (ibuclaw) [inline compiler/test/fail_compilation/fail22384.d]: Does the compiler reject anonymous bitfields with a default member initializer too?

```d
ubyte : 8 = 4;
```

File: compiler/src/dmd/parse.d [hunk 12]
```diff
@@ -4792,11 +4792,9 @@ class Parser(AST, Lexer = dmd.lexer.Lexer) : Lexer
                 AST.Dsymbol s;
                 if (width)
                 {
-                    if (_init)
-                        error("initializer not allowed for bitfield declaration");
                     if (storage_class)
                         error("storage class not allowed for bitfield declaration");
-                    s = new AST.BitFieldDeclaration(loc, t, ident, width);
+                    s = new AST.BitFieldDeclaration(width.loc, t, ident, width, _init);
```
Comment (ibuclaw) [inline compiler/src/dmd/parse.d]: Please rebase against current origin/master.
```suggestion
                    s = new AST.BitFieldDeclaration(loc, t, ident, width, _init);
```

**Comment by ibuclaw** [comment]
Front-end part LGTM. :+1:

File: compiler/test/fail_compilation/fail22384.d (hunk 11 above)
Comment (MadhurKumar004) [reply compiler/test/fail_compilation/fail22384.d]: Yes it does

File: compiler/src/dmd/parse.d (hunk 12 above)
Comment (MadhurKumar004) [reply compiler/src/dmd/parse.d]: Fixed.

File: changelog/dmd.bitfield-default-init.dd (hunk 10 above)
Comment (MadhurKumar004) [reply changelog/dmd.bitfield-default-init.dd]: Fixed.

File: compiler/src/dmd/dsymbolsem.d [hunk 13]
```diff
@@ -3174,6 +3174,12 @@ private extern(C++) final class DsymbolSemanticVisitor : Visitor
         if (dsym.semanticRun >= PASS.semanticdone)
             return;
 
+        if (dsym.isAnonymous() && dsym._init)
+        {
+             .error(dsym.loc, "anonymous bitfield cannot be initialized");
```
Comment (ibuclaw) [inline compiler/src/dmd/dsymbolsem.d]:
```suggestion
             .error(dsym.loc, "anonymous bitfield cannot have default initializer");
```

File: compiler/src/dmd/glue/todt.d [hunk 14]
```diff
@@ -1002,13 +1002,32 @@ private void membersToDt(AggregateDeclaration ad, ref DtBuilder dtb,
...
+        else if (bf)
+        {
+            if (Initializer init = vd._init)
+            {
+                if (init.isVoidInitializer())
+                    continue;
+
+                assert(vd.semanticRun >= PASS.semantic2done);
+                auto ei = init.isExpInitializer();
+                assert(ei);
```
Comment (ibuclaw) [inline compiler/src/dmd/glue/todt.d]:
```suggestion
                assert(ei);
```

File: compiler/test/fail_compilation/fail22384.d [hunk 15]
```diff
@@ -0,0 +1,48 @@
...
+
+enum E : int { A = 1, B = 2 }
+struct EFail {
+    E b : 2 = E.B;
+}
+
+struct IFail {
+    int x : 3 = 4;
+    int y : 7 = 65; 
+    int z : 0 = 1;
```
Comment (ibuclaw) [inline compiler/test/fail_compilation/fail22384.d]:
```suggestion
    int y : 7 = 65;
    int z : 0 = 1;
```

File: compiler/test/fail_compilation/fail22384.d [hunk 16]
```diff
@@ -0,0 +1,48 @@
...
+struct IFail {
+    int x : 3 = 4;
+    int y : 7 = 65; 
+    int z : 0 = 1;  
+}
+
+struct FloatFail {
+    float f : 0;        
+    float f2 : 7;       
+    float f3 : 7 = 4;
```
Comment (ibuclaw) [inline compiler/test/fail_compilation/fail22384.d]:
```suggestion
    float f : 0;
    float f2 : 7;
    float f3 : 7 = 4;
```

File: compiler/test/fail_compilation/fail22384.d [hunk 17]
```diff
@@ -0,0 +1,48 @@
...
+    float f3 : 7 = 4;   
+    float f4 : 7 = 4.2f;
+}
+
+struct InitFail {
+    int i : 7 = 4.2f;   
+    // ubyte : 0 = 4;   
+    ubyte : 8 = 4.2f;   
+    int j : 7 = 'A';    
+    bool b : 7 = 42;
```
Comment (ibuclaw) [inline compiler/test/fail_compilation/fail22384.d]:
```suggestion
    int i : 7 = 4.2f;
    // ubyte : 0 = 4;
    ubyte : 8 = 4.2f;
    int j : 7 = 'A';
    bool b : 7 = 42;
```

**Review by ibuclaw — APPROVED** [review]
Mind there are testsuite failures.
```
==============================
Test 'fail_compilation\biterrors.d' failed: 
diff:
----
-fail_compilation\biterrors.d(103): Error: initializer not allowed for bitfield declaration
 fail_compilation\biterrors.d(104): Error: storage class not allowed for bitfield declaration
----
```

```
==============================
Test 'fail_compilation\biterrors2.d' failed: 
diff:
----
 fail_compilation\biterrors2.d(100): Error: variable `biterrors2.a` - bitfield must be member of struct, union, or class
 int a : 2;
-    ^
+        ^
 fail_compilation\biterrors2.d(104): Error: bitfield `b` has zero width
     int b:0;
           ^
 fail_compilation\biterrors2.d(105): Error: bitfield type `float` is not an integer type
     float c:3;
-          ^
+            ^
----
```

```
==============================
Test 'fail_compilation\fail22384.d' failed: 
diff:
----
-fail_compilation\fail22384.d(21): Error: bitfield initializer `4294967295u` does not fit in 4 bits
-fail_compilation\fail22384.d(26): Error: bitfield initializer `E.B` does not fit in 2 bits
-fail_compilation\fail22384.d(30): Error: bitfield initializer `4` does not fit in 3 bits
-fail_compilation\fail22384.d(31): Error: bitfield initializer `65` does not fit in 7 bits
 fail_compilation\fail22384.d(32): Error: bitfield `z` has zero width
 fail_compilation\fail22384.d(36): Error: bitfield type `float` is not an integer type
 fail_compilation\fail22384.d(36): Error: bitfield `f` has zero width
 fail_compilation\fail22384.d(37): Error: bitfield type `float` is not an integer type
 fail_compilation\fail22384.d(38): Error: bitfield type `float` is not an integer type
 fail_compilation\fail22384.d(39): Error: bitfield type `float` is not an integer type
-fail_compilation\fail22384.d(43): Error: cannot implicitly convert expression `4.2F` of type `float` to `int`
 fail_compilation\fail22384.d(45): Error: anonymous bitfield cannot be initialized
+fail_compilation\fail22384.d(21): Error: bitfield initializer `4294967295u` does not fit in 4 bits
+fail_compilation\fail22384.d(26): Error: bitfield initializer `E.B` does not fit in 2 bits
+fail_compilation\fail22384.d(30): Error: bitfield initializer `4` does not fit in 3 bits
+fail_compilation\fail22384.d(31): Error: bitfield initializer `65` does not fit in 7 bits
+fail_compilation\fail22384.d(43): Error: cannot implicitly convert expression `4.2F` of type `float` to `int`
 fail_compilation\fail22384.d(46): Error: bitfield initializer `65` does not fit in 7 bits
 fail_compilation\fail22384.d(47): Error: cannot implicitly convert expression `42` of type `int` to `bool`
----
```

File: compiler/src/dmd/parse.d (hunk 12 above)
Comment (MadhurKumar004) [reply compiler/src/dmd/parse.d]: Can you please expand on what you mean by Location is still wrong?

**Comment by MadhurKumar004** [comment]
@ibuclaw fixed the testsuite failure , whitespace and the correct error message. Also can you please explain what you mean by location is still wrong?

**Comment by MadhurKumar004** [comment]
@ibuclaw any update?

**Comment by ibuclaw** [comment]
Fixed merge conflicts, waiting for test results.

**Comment by ibuclaw** [comment]
```
src/dmd/parse.d(4797): Error: constructor `dmd.astbase.ASTBase.BitFieldDeclaration.this(Loc loc, Type type, Identifier id, Expression width)` is not callable using argument types `(const(Loc), Type, Identifier, Expression, Initializer)`
                    s = new AST.BitFieldDeclaration(loc, t, ident, width, _init);
                        ^
```
Ugh, need to update the dummy AST definitions in [compiler/src/dmd/astbase.d](https://github.com/dlang/dmd/blob/master/compiler/src/dmd/astbase.d#L542).

File: compiler/src/dmd/astbase.d [hunk 18]
```diff
@@ -539,10 +539,9 @@ struct ASTBase
         uint fieldWidth;
         uint bitOffset;
 
-        final extern (D) this(Loc loc, Type type, Identifier id, Expression width)
+        final extern (D) this(Loc loc, Type type, Identifier id, Expression width, Initializer _inti = null)
```
Comment (ibuclaw) [inline compiler/src/dmd/astbase.d]: Typo.
```suggestion
        final extern (D) this(Loc loc, Type type, Identifier id, Expression width, Initializer _init = null)
```

**Comment by ibuclaw** [comment]
@MadhurKumar004 can you rebase against current dmd/master? This branch still has conflicts.

File: compiler/test/fail_compilation/fail22384.d [hunk 19]
```diff
@@ -0,0 +1,47 @@
...
+fail_compilation/fail22384.d(32): Error: bitfield `z` has zero width
+fail_compilation/fail22384.d(36): Error: bitfield type `float` is not an integer type
+fail_compilation/fail22384.d(36): Error: bitfield `f` has zero width
+fail_compilation/fail22384.d(37): Error: bitfield type `float` is not an integer type
+fail_compilation/fail22384.d(38): Error: bitfield type `float` is not an integer type
+fail_compilation/fail22384.d(39): Error: bitfield type `float` is not an integer type
+fail_compilation/fail22384.d(44): Error: anonymous bitfield cannot have default initializer
+fail_compilation/fail22384.d(21): Error: bitfield initializer `4294967295u` does not fit in 4 bits
+fail_compilation/fail22384.d(26): Error: bitfield initializer `E.B` does not fit in 2 bits
+fail_compilation/fail22384.d(30): Error: bitfield initializer `4` does not fit in 3 bits
```
Comment (ibuclaw) [inline compiler/test/fail_compilation/fail22384.d:13]: This could be clarified that it's a signed 3-bit type, as `0b100` _is_ three bits.

But, OK to defer to a follow-up PR.

**Review by ibuclaw — APPROVED** [review]

**Comment by ibuclaw** [comment]
Front-end parts LGTM, again can't comment on back-end.

**Comment by MadhurKumar004** [comment]
Should i wait for someone to review back-end?

**Comment by ibuclaw** [comment]
I'll give it ~72h, unless someone merges sooner.

Thanks! 🙇

**Comment by thewilsonator** [comment]
> I'll give it ~72h, unless someone merges sooner.

I would, but it has merge conflicts.

**Comment by ibuclaw** [comment]
> > I'll give it ~72h, unless someone merges sooner.
> 
> I would, but it has merge conflicts.
> 

Squash and merge?

**Comment by thewilsonator** [comment]
> This branch cannot be rebased due to conflicts

**Comment by MadhurKumar004** [comment]
I dont know why this keeps happening, there was just one merge conflict and that too i resolved after pushing.

**Comment by MadhurKumar004** [comment]
Should i push it again after rebasing?

**Comment by thewilsonator** [comment]
yes

**Comment by MadhurKumar004** [comment]
@ibuclaw @thewilsonator A bit late, but thank you very much for your time and for reviewing this pr. I appreciate your help and suggestion!. Let me know if there is anything else to be addressed.
